pkg/sdkserver/sdkserver.go

```diff
@@ -762,20 +789,190 @@ func (s *SDKServer) GetCounter(ctx context.Context, in *alpha.GetCounterRequest)
+		countSet := in.CounterUpdateRequest.Count.GetValue()
+		capacity := batchCounter.counter.Capacity
+		if batchCounter.capacitySet != nil {
+			capacity = *batchCounter.capacitySet
+		}
+		if countSet < 0 || countSet > capacity {
+			return nil, errors.Errorf("OUT_OF_RANGE. Count must be within range [0,Capacity]. Found Count: %d, Capacity: %d", countSet, capacity)
+		}
+		batchCounter.countSet = &countSet
+		// Clear any previous CountIncrement or CountDecrement requests, and add the CountSet as the first item.
+		batchCounter.diff = 0
+	case in.CounterUpdateRequest.Capacity != nil: // Updated based on if client call is CapacitySet
+		if in.CounterUpdateRequest.Capacity.GetValue() < 0 {
+			return nil, errors.Errorf("OUT_OF_RANGE. Capacity must be greater than or equal to 0. Found Capacity: %d", in.CounterUpdateRequest.Capacity.GetValue())
+		}
+		capacitySet := in.CounterUpdateRequest.Capacity.GetValue()
+		batchCounter.capacitySet = &capacitySet
+	default:
+		return nil, errors.Errorf("INVALID_ARGUMENT. Malformed CounterUpdateRequest: %v", in.CounterUpdateRequest)
+	}
+
+	s.gsCounterUpdates[name] = batchCounter
+
+	// Queue up the Update for later batch processing by updateCounters.
+	s.workerqueue.Enqueue(cache.ExplicitKey(updateCounters))
+	return &alpha.Counter{}, nil
+}
+
+// updateCounter updates the Counters in the GameServer's Status with the batched update requests.
+func (s *SDKServer) updateCounter(ctx context.Context) error {
+	gs, err := s.gameServer()
+	if err != nil {
+		return err
+	}
+	gsCopy := gs.DeepCopy()
+
+	s.logger.WithField("batchCounterUpdates", s.gsCounterUpdates).Debug("Batch updating Counter(s)")
+	s.gsUpdateMutex.Lock()
+	defer s.gsUpdateMutex.Unlock()
+
+	names := []string{}
+
+	for name, ctrReq := range s.gsCounterUpdates {
+		counter, ok := gsCopy.Status.Counters[name]
+		if !ok {
+			continue
+		}
+		// Changes may have been made to the Counter since we validated the incoming changes in
+		// UpdateCounter, and we need to verify if the batched changes can be fully applied, partially
+		// applied, or cannot be applied.
+		if ctrReq.capacitySet != nil {
+			counter.Capacity = *ctrReq.capacitySet
+		}
+		if ctrReq.countSet != nil {
+			counter.Count = *ctrReq.countSet
+		}
+		newCnt := counter.Count + ctrReq.diff
+		if newCnt < 0 {
+			newCnt = 0
+			s.logger.Debug("truncating Count in Update Counter request to 0")
+		}
+		if newCnt > counter.Capacity {
+			newCnt = counter.Capacity
+			s.logger.Debug("truncating Count in Update Counter request to Capacity")
+		}
+		counter.Count = newCnt
+		gsCopy.Status.Counters[name] = counter
+		names = append(names, name)
+	}
+
+	gs, err = s.gameServerGetter.GameServers(s.namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Record an event per update Counter
+	for _, name := range names {
+		s.recorder.Event(gs, corev1.EventTypeNormal, "UpdateCounter",
+			fmt.Sprintf("Counter %s updated", name))
```

Comment:
I really liked what you had with the details of what was updated for each counter in the event - it is possible to still capture that here?

Reply:
Yep, I can add that back in.